plugins/inputs/dpdk/README.md

```diff
@@ -0,0 +1,184 @@
+examples for further details.
+
+### Example: Independent NIC link status configuration
+This configuration allows getting `/ethdev/link_status` using separate configuration, with higher timeout.
+```toml
+[[inputs.dpdk]]
+  interval = "30s"
+  socket_access_timeout = "10s"
+  device_types = ["ethdev"]
+
+  [inputs.dpdk.ethdev]
+    exclude_commands = ["/ethdev/stats", "/ethdev/xstats"]
+```
+
+### Example: Getting application-specific metrics
+This configuration allows reading custom metrics exposed by applications. Example telemetry command obtained from  
+[L3 Forwarding with Power Management Sample Application](https://doc.dpdk.org/guides/sample_app_ug/l3_forward_power_man.html).
+```toml
+[[inputs.dpdk]]
+  device_types = ["ethdev"]
+  additional_commands = ["/l3fwd-power/stats"]
+
+  [inputs.dpdk.ethdev]
+    exclude_commands = ["/ethdev/link_status"]
+```
+Command entries specified in `additional_commands` should match DPDK command format:
+* Command entry format: either `command` or `command,params` for commands that expect parameters, where comma (`,`) separates command from params.
+* Command entry length (command with params) should be `< 1024` characters.
+* Command length (without params) should be `< 56` characters.
+* Commands have to start with `/`.
+
+Providing invalid commands will prevent the plugin from starting. Additional commands allow duplicates, but they
+will be removed during execution so each command will be executed only once during each metric gathering interval. 
+
+### Example: Getting metrics from multiple DPDK instances running on same host
+This configuration allows getting metrics from two separate applications exposing their telemetry interfaces
+via separate sockets. For each plugin instance a unique tag `[inputs.dpdk.tags]` allows distinguishing between them. 
+```toml
+# Instance #1 - L3 Forwarding with Power Management Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l3fwd-power_telemetry.v2"
+  device_types = ["ethdev"]
+  additional_commands = ["/l3fwd-power/stats"]
+
+  [inputs.dpdk.ethdev]
+    exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l3fwd-power"
+
+# Instance #2 - L2 Forwarding with Intel Cache Allocation Technology (CAT) Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l2fwd-cat_telemetry.v2"
+  device_types = ["ethdev"]
+
+[inputs.dpdk.ethdev]
+  exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l2fwd-cat"
+```
+This utilizes Telegraf's standard capability of [adding custom tags](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#input-plugins)
+to input plugin's measurements.
+
+## Metrics
+The DPDK socket accepts `command,params` requests and returns metric data in JSON format. All metrics from DPDK socket
+become flattened using [Telegraf's JSON Flattener](../../parsers/json/README.md) and exposed as fields. 
```

Comment:
I have concerns about the metric format in this plugin.  I don't think I like the field names that the json flattener produces.  It's not user friendly to have 80-200 fields with very similar names.  I'm not sure exactly what to do here because I'm not an expert on DPDK, but having so many fields with various prefixes and suffixes doesn't feel right.

Maybe that prefix/suffix information should be exposed as tags?  Something like direction='tx' or 'rx' would cut down on the number of fields in each metric.  (maybe direction isn't the right tag name, but you get the idea)  I see lots of names with i or o, presumably input and output?  It seems like that should be a tag too.

I also see histograms by packet size in the fields, like "rx_size_65_to_127_packets".  Would it make sense to output these more like the histogram aggregator? (https://github.com/influxdata/telegraf/tree/master/plugins/aggregators/histogram)  I could also see people using the prometheus output wanting the size histograms to work with prometheus's histogram metric type.

Reply:
Especially for applications that may expose their own metrics through DPDK telemetry interface we have no insights into the structure of those responses since they may contain completely custom metrics. If we want to expose all metrics, even the ones that are available in some nested structures (that we don't know the format of), then those not-so-friendly names are a cost of just having those metrics even collected.

Considering the set of metrics exposed in the example output provided in plugin readme: those come from an Intel driver that exposes quite a bit of detailed NIC metrics, including the packet size buckets that you mentioned. Unfortunately here we also don't have a common naming standard that we can rely on, that is available across all drivers. It means we would not be able to clearly interpret them and e.g. break down into groups of metrics sharing a common set of tags. 

Due to such per-driver, per-vendor and per-application differences in metrics we went with the approach of reducing transformation of metrics to a minimum that is necessary to expose them, and to just provide a set of tags to enhance querying capabilities (host, dpdk_instance, command, param).

Reply:
Since DPDK libraries and applications don't provide a standardized format, it's understandable that DPDK has to provide a generic metric format to telegraf. I'm ok accepting the generic metric format, but I think this should be an area we to try to improve in the future. Maybe with improvements to json support in telegraf we will be able to come up with a way to use tags and/or metric names instead of field prefixes.

Reply:
No changes here right now, as we agreed.